Import the existing bioblend tutorial, exercises (#3505)
galaxyproject/training-material: topics/dev/tutorials/bioblend-api/tutorial.md

---
layout: tutorial_hands_on

title: "Scripting Galaxy using the API and BioBlend"
level: Introductory
requirements: []
follow_up_training: []
questions:
  - What is a REST API?
  - How to interact with Galaxy programmatically?
  - Why and when should I use BioBlend?

objectives:
  - Interact with Galaxy via BioBlend.

time_estimation:  2h
key_points:
  - The API allows you to use Galaxy's capabilities programmatically.
  - BioBlend makes using the Galaxy API from Python easier.
  - BioBlend objects is an object-oriented interface for interacting with Galaxy.

subtopic: api
contributions:
  authorship:
  - nsoranzo
  editing:
  - claresloggett
  - nturaga
  - hexylena

notebook:
  language: python
---

> ### Agenda
>
> In this tutorial, we will cover:
>
> 1. TOC
> {:toc}
>
{: .agenda}

# Interacting with histories in Galaxy API

> We are going to use the [requests](http://python-requests.org/) Python library to communicate via HTTP with the Galaxy server. To start, let's define the connection parameters.

Reviewer: ```suggestion
We are going to use the [requests](https://requests.readthedocs.io/) Python library to communicate via HTTP with the Galaxy server. To start, let's define the connection parameters.
```

**You need to insert the API key for your Galaxy server in the cell below**:
1. Open the Galaxy server in another browser tab
2. Click on "User" on the top menu, then "Preferences"
3. Click on "Manage API key"
4. Generate an API key if needed, then copy the alphanumeric string and paste it as the value of the `api_key` variable below.

```python

> from pprint import pprint
> 
> import json
> 
> import requests
> from six.moves.urllib.parse import urljoin

Reviewer: ```suggestion
import json
from pprint import pprint
from urllib.parse import urljoin

import requests
```

server = 'https://usegalaxy.eu/'
api_key = ''
base_url = urljoin(server, 'api')
base_url
```

We now make a GET request to retrieve all histories owned by a user:

```python
params = {'key': api_key}

> r = requests.get(base_url + '/histories', params)

Author: wow, so many small changes have happened over the years. Glad you caught these.

- `state_ids`: ids of history datasets for each possible state.

To get the list of **datasets contained** in a history, simply append `/contents` to the previous resource request.

```python
params = {'key': api_key}
r = requests.get(base_url + '/histories/' + hist0_id + '/contents', params)
hdas = r.json()
pprint(hdas)
```

The dictionaries returned when GETting the history content give basic info about each dataset, e.g.: `id`, `name`, `deleted`, `state`, `url`...

To get the details about a specific dataset, you can use the `datasets` controller:

```python
hda0_id = hdas[0]['id']
print(hda0_id)
params = {'key': api_key}
r = requests.get(base_url + '/datasets/' + hda0_id, params)
pprint(r.json())
```

Some of the interesting additional dictionary entries are:
- `create_time`
- `creating job`: id of the job which created this dataset
- `download_url`: URL to download the dataset
- `file_ext`: the Galaxy data type of this dataset
- `file_size`
- `genome_build`: the genome build (dbkey) associated to this dataset.

**New resources** are created with POST requests. The uploaded **data needs to be serialized** in a JSON string. For example, to create a new history:

```python

> params = {'key': api_key}

Reviewer: ```suggestion
```

data = {'name': 'New history'}

> r = requests.post(base_url + '/histories', data=json.dumps(data), params=params, headers={'Content-Type': 'application/json'})

Reviewer: ```suggestion
r = requests.post(base_url + "/histories", data=json.dumps(data), headers=headers)
```

new_hist = r.json()
pprint(new_hist)
```

The return value of a POST request is a dictionary with detailed info about the created resource.

To **update** a resource, make a PUT request, e.g. to change the history name:

```python

> params = {'key': api_key}

Reviewer: ```suggestion
```

data = {'name': 'Updated history'}

> r = requests.put(base_url + '/histories/' + new_hist['id'], json.dumps(data), params=params, headers={'Content-Type': 'application/json'})

Reviewer: ```suggestion
r = requests.put(base_url + "/histories/" + new_hist["id"], json.dumps(data), headers=headers)
```

print(r.status_code)
pprint(r.json())
```

The return value of a PUT request is usually a dictionary with detailed info about the updated resource.

Finally to **delete** a resource, make a DELETE request, e.g.:

```python

> params = {'key': api_key}
> r = requests.delete(base_url + '/histories/' + new_hist['id'], params=params)

Reviewer: ```suggestion
r = requests.delete(base_url + "/histories/" + new_hist["id"], headers=headers)
```

print(r.status_code)
```

## Exercise: Galaxy API

**Goal**: Upload a file to a new history, import a workflow and run it on the uploaded dataset.

> ### {% icon question %} Question: Initialise
> First, define the connection parameters. What variables do you need?
> > ### {% icon solution %} Solution
> > ```python

> > > from pprint import pprint
> > >
> > > import json
> > >
> > > import requests
> > > from six.moves.urllib.parse import urljoin

Reviewer: ```suggestion
> > import json
> > from pprint import pprint
> > from urllib.parse import urljoin
> >
> > import requests
```

> >
> > server = 'https://usegalaxy.eu/'
> > api_key = ''
> > base_url = urljoin(server, 'api')
> > ```
> >
> {: .solution}
{: .question}

```python
# Try it out here!

```

> ### {% icon question %} Question: New History
> Next, create a new Galaxy history via POST to the correct API.
> > ### {% icon solution %} Solution
> > ```python

> > > params = {'key': api_key}

Reviewer: ```suggestion
> > headers = {"Content-Type": "application/json", "x-api-key": api_key}

```

> > data = {'name': 'New history'}

> > > r = requests.post(base_url + '/histories', json.dumps(data), params=params, headers={'Content-Type': 'application/json'})

Reviewer: ```suggestion
> > r = requests.post(base_url + "/histories", json.dumps(data), headers=headers)
```

> > ret = r.json()
> > pprint(ret)
> > ```
> >
> {: .solution}
{: .question}

```python
# Try it out here!

```

> ### {% icon question %} Question: Find the dataset in your history
> Find the new uploaded dataset, either from the dict returned by the POST request or from the history contents.
> > ### {% icon solution %} Solution
> > ```python
> > hda = ret['outputs'][0]
> > pprint(hda)
> > ```
> >
> {: .solution}
{: .question}

```python
# Try it out here!

```

> ### {% icon question %} Question: Import a workflow
> **Import a workflow** from the local file "test-data/convert_to_tab.ga" by making a `POST` request to `/api/workflows`. The only needed data is 'workflow', which must be a deserialized JSON representation of the workflow.
>
> You can obtain this file from the following URL, you'll need to download it first.
>
> ```
> https://raw.githubusercontent.com/nsoranzo/bioblend-tutorial/main/test-data/convert_to_tab.ga
> ```
> > ### {% icon solution %} Solution
> >
> > ```python
> > params = {'key': api_key}
> > with open('test-data/convert_to_tab.ga', 'r') as f:
> >     workflow_json = json.load(f)
> > data = {'workflow': workflow_json}

> > > r = requests.post(base_url + '/workflows', json.dumps(data), params=params, headers={'Content-Type': 'application/json'})

Reviewer: ```suggestion
> > r = requests.post(base_url + "/workflows", json.dumps(data), headers=headers)
```

> > wf = r.json()
> > pprint(wf)
> > ```
> {: .solution}
{: .question}

```python
# Try it out here!

```

> ### {% icon question %} Question: View the WF details
> View the details of the imported workflow by making a GET request to `/api/workflows`.
> > ### {% icon solution %} Solution
> > ```python

> > > params = {'key': api_key}
> > > r = requests.get(base_url + '/workflows/' + wf['id'], params)

Reviewer: ```suggestion
> > r = requests.get(base_url + "/workflows/" + wf["id"], headers=headers)
```

> > wf = r.json()
> > pprint(wf)
> > ```
> >
> {: .solution}
{: .question}

```python
# Try it out here!

```

> ### {% icon question %} Question: Run the WF
> **Run** the imported workflow on the uploaded dataset **inside the same history** by making a `POST` request to `/api/workflows/<id>/invocations`. The only needed data are 'history' and 'inputs'.
> > ### {% icon solution %} Solution
> >
> > ```python
> > params = {'key': api_key}
> > inputs = {0: {'id': hda['id'], 'src': 'hda'}}
> > data = {
> >     'history': 'hist_id=' + new_hist['id'],
> >     'inputs': inputs}

> > > r = requests.post(base_url + '/workflows/' + wf['id'] + '/invocations', json.dumps(data), params=params, headers={'Content-Type': 'application/json'})

Reviewer: ```suggestion
> > r = requests.post(base_url + "/workflows/" + wf["id"] + "/invocations", json.dumps(data), headers=headers)
```
